// datauri/old/api.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _events = require('events');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _mimer = require('mimer');

var _mimer2 = _interopRequireDefault(_mimer);

var _imageSize = require('image-size');

var _imageSize2 = _interopRequireDefault(_imageSize);

var _uri = require('./template/uri');

var _uri2 = _interopRequireDefault(_uri);

var _css = require('./template/css');

var _css2 = _interopRequireDefault(_css);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var existsSync = _fs2.default.existsSync;
var exists = _fs2.default.exists;

var Api = (function (_EventEmitter) {
  _inherits(Api, _EventEmitter);

  function Api() {
    _classCallCheck(this, Api);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Api).call(this));
  }

  _createClass(Api, [{
    key: 'format',
    value: function format(fileName, fileContent) {
      fileContent = fileContent instanceof Buffer ? fileContent : new Buffer(fileContent);
      this.fileName = fileName;
      this.base64 = fileContent.toString('base64');
      this.mimetype = (0, _mimer2.default)(fileName);
      this.content = (0, _uri2.default)({
        base64: this.base64,
        mimetype: this.mimetype
      });

      return this;
    }
  }, {
    key: 'encode',
    value: function encode(fileName, handler) {
      var _this2 = this;

      return this.async(fileName, function (err) {
        if (handler) {
          if (err) {
            return handler(err);
          }

          handler.call(_this2, null, _this2.content, _this2);

          return _this2;
        }

        if (err) {
          _this2.emit('error', err);

          return _this2;
        }

        _this2.emit('encoded', _this2.content, _this2);
      });
    }
  }, {
    key: 'encodeSync',
    value: function encodeSync(fileName) {
      if (!fileName || !fileName.trim || fileName.trim() === '') {
        throw new Error('Insert a File path as string argument');
      }

      if (existsSync(fileName)) {
        var fileContent = _fs2.default.readFileSync(fileName);

        return this.format(fileName, fileContent).content;
      }

      throw new Error('The file ' + fileName + ' was not found!');
    }
  }, {
    key: 'async',
    value: function async(fileName, handler) {
      var _this3 = this;

      exists(fileName, function () {
        _fs2.default.readFile(fileName, function (err, fileContent) {
          if (err) {
            return handler.call(_this3, err);
          }

          handler.call(_this3.format(fileName, fileContent));
        });
      });
    }
  }, {
    key: 'getCSS',
    value: function getCSS() {
      var config = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

      if (!this.content) {
        throw new Error('Create a data-uri config using the method encodeSync');
      }

      config.class = config.class || _path2.default.basename(this.fileName, _path2.default.extname(this.fileName));
      config.background = this.content;

      if (config.width || config.height || config['background-size']) {
        config.dimensions = (0, _imageSize2.default)(this.fileName);
      }

      return (0, _css2.default)(config);
    }
  }]);

  return Api;
})(_events.EventEmitter);

exports.default = Api;
